In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime

In [2]:
df = pd.read_csv('ConsumerElectronics.csv',na_values=['\\N',''])


C:\Users\rakes\Anaconda_new\lib\site-packages\IPython\core\interactiveshell.py:3057: DtypeWarning: Columns (6,12,13) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df.head(10)

fsn_id           order_date  Year  Month      order_id  \
0  ACCCX3S58G7B5F6P  2015-10-17 15:11:54  2015     10  3.419301e+15   
1  ACCCX3S58G7B5F6P  2015-10-19 10:07:22  2015     10  1.420831e+15   
2  ACCCX3S5AHMF55FV  2015-10-20 15:45:56  2015     10  2.421913e+15   
3  ACCCX3S5AHMF55FV  2015-10-14 12:05:15  2015     10  4.416592e+15   
4  ACCCX3S5AHMF55FV  2015-10-17 21:25:03  2015     10  4.419525e+15   
5  ACCCX3S5JGAJETYR  2015-10-17 12:07:24  2015     10  3.419189e+15   
6  ACCCX3S5JGAJETYR  2015-10-16 14:59:57  2015     10  3.418429e+15   
7  ACCCX3S5JGAJETYR  2015-10-15 23:41:39  2015     10  3.417799e+15   
8  ACCCX3S5JGAJETYR  2015-10-16 00:49:14  2015     10  2.417870e+15   
9  ACCCX3S5JGAJETYR  2015-10-25 22:52:34  2015     10  4.426482e+15   

   order_item_id   gmv  units  deliverybdays  deliverycdays  \
0   3.419301e+15  6400      1            NaN            NaN   
1   1.420831e+15  6900      1            NaN            NaN   
2   2.421913e+15  1990      1            NaN            NaN   
3   4.416592e+15  1690      1            NaN            NaN   
4   4.419525e+15  1618      1            NaN            NaN   
5   3.419189e+15  3324      1            NaN            NaN   
6   3.418429e+15  3695      1            NaN            NaN   
7   3.417799e+15  3695      1            NaN            NaN   
8   2.417870e+15            1            NaN            NaN   
9   4.426482e+15  3695      1            NaN            NaN   

  s1_fact.order_payment_type  sla                 cust_id  \
0                        COD    5  -1.01299130778588E+018   
1                        COD    7  -8.99032457905512E+018   
2                        COD   10   -1.0404429420466E+018   
3                    Prepaid    4  -7.60496084352714E+018   
4                    Prepaid    6    2.8945572083453E+018   
5                    Prepaid    5  -7.64154636189174E+018   
6                    Prepaid    6  -7.15614041481066E+016   
7                        COD    5   9.17327596234152E+018   
8                    Prepaid    9                           
9                        COD    7  -7.80915606896163E+018   

                  pincode product_analytic_super_category  \
0  -7.79175582905735E+018                              CE   
1   7.33541149097431E+018                              CE   
2  -7.47768776228657E+018                              CE   
3  -5.83593163877661E+018                              CE   
4   5.34735360997242E+017                              CE   
5  -1.91905269756031E+018                              CE   
6   6.40491320207757E+018                              CE   
7  -1.60630305201981E+018                              CE   
8                                                      CE   
9    -9.779010321304E+017                              CE   

  product_analytic_category product_analytic_sub_category  \
0           CameraAccessory               CameraAccessory   
1           CameraAccessory               CameraAccessory   
2           CameraAccessory               CameraAccessory   
3           CameraAccessory               CameraAccessory   
4           CameraAccessory               CameraAccessory   
5           CameraAccessory               CameraAccessory   
6           CameraAccessory               CameraAccessory   
7           CameraAccessory               CameraAccessory   
8           CameraAccessory               CameraAccessory   
9           CameraAccessory               CameraAccessory   

  product_analytic_vertical  product_mrp  product_procurement_sla  
0              CameraTripod         7190                        0  
1              CameraTripod         7190                        0  
2              CameraTripod         2099                        3  
3              CameraTripod         2099                        3  
4              CameraTripod         2099                        3  
5              CameraTripod         4044                        5  
6              CameraTripod         4044     

In [4]:
df.shape

(1648824, 20)

### Data preparation

In [5]:
### Converting gmv, cust_id, pincode to numeric to read the null values

df.gmv = pd.to_numeric(df.gmv, errors="coerce")
df.pincode = pd.to_numeric(df.pincode, errors="coerce")
df.cust_id = pd.to_numeric(df.cust_id, errors="coerce")

In [6]:
# null values checks
df.isnull().sum()

fsn_id                                   0
order_date                               0
Year                                     0
Month                                    0
order_id                                 0
order_item_id                            0
gmv                                   4904
units                                    0
deliverybdays                      1312972
deliverycdays                      1312971
s1_fact.order_payment_type               0
sla                                      0
cust_id                               4904
pincode                               4904
product_analytic_super_category          0
product_analytic_category                0
product_analytic_sub_category            0
product_analytic_vertical             5828
product_mrp                              0
product_procurement_sla                  0
dtype: int64

There are 4904 missing values for gmv,pincode and cust_id

In [7]:
# To remove null values from gmv,pincode and cust_id
df=df[~np.isnan(df['gmv'])]
df=df[~np.isnan(df['cust_id'])]
df=df[~np.isnan(df['pincode'])]

In [8]:
# column-wise null percentages here
round(100*(df.isnull().sum()/len(df.index)),2)

fsn_id                              0.00
order_date                          0.00
Year                                0.00
Month                               0.00
order_id                            0.00
order_item_id                       0.00
gmv                                 0.00
units                               0.00
deliverybdays                      79.57
deliverycdays                      79.57
s1_fact.order_payment_type          0.00
sla                                 0.00
cust_id                             0.00
pincode                             0.00
product_analytic_super_category     0.00
product_analytic_category           0.00
product_analytic_sub_category       0.00
product_analytic_vertical           0.35
product_mrp                         0.00
product_procurement_sla             0.00
dtype: float64

In [9]:
# Dropping deliverybdays and deliverycdays  column since it's having 80% null values.
df=df.drop(['deliverybdays','deliverycdays'],axis=1)

In [10]:
df.shape

(1643920, 18)

In [11]:
df=df.dropna(subset=['product_analytic_vertical'],axis=0)

In [12]:
df.isnull().sum()

fsn_id                             0
order_date                         0
Year                               0
Month                              0
order_id                           0
order_item_id                      0
gmv                                0
units                              0
s1_fact.order_payment_type         0
sla                                0
cust_id                            0
pincode                            0
product_analytic_super_category    0
product_analytic_category          0
product_analytic_sub_category      0
product_analytic_vertical          0
product_mrp                        0
product_procurement_sla            0
dtype: int64

In [13]:
df.shape

(1638106, 18)

In [14]:
# Number of rows with MRP value as zero
df[df['product_mrp']==0].shape


(5290, 18)

We have 5290 records with product_mrp as zero value, which is not a valid scenario. We cannot have product with product_mrp as zero. Hence deleting those recors

In [15]:
# Delete product with product_mrp as zero
df=df[df.product_mrp!=0]

In [16]:
df.shape

(1632816, 18)

In [17]:
# Number of records with product_procurement_sla as -1
df[df['product_procurement_sla']==-1].shape

(70321, 18)

Negative values in product_procurement_sla is invalid scenario. Hence replacing with zero values

In [18]:
# replace product_procurement_sla negative values with zero
df['product_procurement_sla']=df['product_procurement_sla'].replace(-1,0)

In [19]:
df.shape

(1632816, 18)

In [20]:
# Almost 50% cust_id is having negative values
df[df['cust_id']<0].shape

(817926, 18)

In [21]:
# Almost 50% pincode is having negative values
df[df['pincode']<0].shape

(805932, 18)

In [22]:
# Dropping cust_id and pincode from dataset since it will not helpful for the model
df=df.drop(['cust_id','pincode'],axis=1)


In [23]:
df.describe()

Year         Month      order_id  order_item_id           gmv  \
count  1.632816e+06  1.632816e+06  1.632816e+06   1.632816e+06  1.632816e+06   
mean   2.015519e+03  6.582622e+00  2.842424e+15   1.371587e+16  2.450530e+03   
std    4.996213e-01  3.625040e+00  1.268668e+15   5.584622e+16  5.605788e+03   
min    2.015000e+03  1.000000e+00  7.633152e+07   1.080165e+08  0.000000e+00   
25%    2.015000e+03  3.000000e+00  1.602417e+15   1.606559e+15  3.400000e+02   
50%    2.016000e+03  6.000000e+00  2.611734e+15   2.619034e+15  7.490000e+02   
75%    2.016000e+03  1.000000e+01  3.610817e+15   4.397626e+15  1.999000e+03   
max    2.016000e+03  1.200000e+01  5.395077e+15   3.000373e+17  2.269470e+05   

              units           sla   product_mrp  product_procurement_sla  
count  1.632816e+06  1.632816e+06  1.632816e+06             1.632816e+06  
mean   1.021943e+00  5.693237e+00  4.103883e+03             5.482460e+00  
std    2.518061e-01  2.984569e+00  8.528829e+03             5.366434e+01  
min    1.000000e+00  0.000000e+00  4.900000e+01             0.000000e+00  
25%    1.000000e+00  4.000000e+00  7.990000e+02             1.000000e+00  
50%    1.000000e+00  6.000000e+00  1.599000e+03             2.000000e+00  
75%    1.000000e+00  7.000000e+00  3.495000e+03             3.000000e+00  
max    5.000000e+01  1.006000e+03  2.999990e+05             1.000000e+03

In [24]:
df.head()

fsn_id           order_date  Year  Month      order_id  \
0  ACCCX3S58G7B5F6P  2015-10-17 15:11:54  2015     10  3.419301e+15   
1  ACCCX3S58G7B5F6P  2015-10-19 10:07:22  2015     10  1.420831e+15   
2  ACCCX3S5AHMF55FV  2015-10-20 15:45:56  2015     10  2.421913e+15   
3  ACCCX3S5AHMF55FV  2015-10-14 12:05:15  2015     10  4.416592e+15   
4  ACCCX3S5AHMF55FV  2015-10-17 21:25:03  2015     10  4.419525e+15   

   order_item_id     gmv  units s1_fact.order_payment_type  sla  \
0   3.419301e+15  6400.0      1                        COD    5   
1   1.420831e+15  6900.0      1                        COD    7   
2   2.421913e+15  1990.0      1                        COD   10   
3   4.416592e+15  1690.0      1                    Prepaid    4   
4   4.419525e+15  1618.0      1                    Prepaid    6   

  product_analytic_super_category product_analytic_category  \
0                              CE           CameraAccessory   
1                              CE           CameraAccessory   
2                              CE           CameraAccessory   
3                              CE           CameraAccessory   
4                              CE           CameraAccessory   

  product_analytic_sub_category product_analytic_vertical  product_mrp  \
0               CameraAccessory              CameraTripod         7190   
1               CameraAccessory              CameraTripod         7190   
2               CameraAccessory              CameraTripod         2099   
3               CameraAccessory              CameraTripod         2099   
4               CameraAccessory              CameraTripod         2099   

   product_procurement_sla  
0                        0  
1                        0  
2                        3  
3                        3  
4                        3

Adding new column Week_Number

In [25]:
# convert order_date to datetime
df['order_date']=pd.to_datetime(df['order_date'], dayfirst=True)

In [26]:
# Extract week number from order_date
df['Week_Number'] = df['order_date'].dt.week

Adding new column pay_flag

In [27]:
# Add pay_flag
df['pay_flag']= df.order_date.dt.day.isin([1,15]).astype(np.int8)

Adding new column Holiday_flag


In [28]:
df['order_date']=df['order_date'].dt.strftime('%Y-%m-%d')

In [29]:
df.head()

fsn_id  order_date  Year  Month      order_id  order_item_id  \
0  ACCCX3S58G7B5F6P  2015-10-17  2015     10  3.419301e+15   3.419301e+15   
1  ACCCX3S58G7B5F6P  2015-10-19  2015     10  1.420831e+15   1.420831e+15   
2  ACCCX3S5AHMF55FV  2015-10-20  2015     10  2.421913e+15   2.421913e+15   
3  ACCCX3S5AHMF55FV  2015-10-14  2015     10  4.416592e+15   4.416592e+15   
4  ACCCX3S5AHMF55FV  2015-10-17  2015     10  4.419525e+15   4.419525e+15   

      gmv  units s1_fact.order_payment_type  sla  \
0  6400.0      1                        COD    5   
1  6900.0      1                        COD    7   
2  1990.0      1                        COD   10   
3  1690.0      1                    Prepaid    4   
4  1618.0      1                    Prepaid    6   

  product_analytic_super_category product_analytic_category  \
0                              CE           CameraAccessory   
1                              CE           CameraAccessory   
2                              CE           CameraAccessory   
3                              CE           CameraAccessory   
4                              CE           CameraAccessory   

  product_analytic_sub_category product_analytic_vertical  product_mrp  \
0               CameraAccessory              CameraTripod         7190   
1               CameraAccessory              CameraTripod         7190   
2               CameraAccessory              CameraTripod         2099   
3               CameraAccessory              CameraTripod         2099   
4               CameraAccessory              CameraTripod         2099   

   product_procurement_sla  Week_Number  pay_flag  
0                        0           42         0  
1                        0           43         0  
2                        3           43         0  
3                        3           42         0  
4                        3           42         0

Adding new column Holiday_flag

In [30]:
holiday={'holiday':['2015-06-18','2015-06-19','2015-08-15','2015-08-16','2015-08-17','2015-08-28','2015-08-29','2015-08-30','2015-10-17','2015-10-16','2015-10-15','2015-11-07','2015-11-08','2015-11-09','2015-11-10','2015-11-11','2015-11-12','2015-11-13','2015-11-14','2015-12-25','2015-12-26','2015-12-27','2015-12-28','2015-12-29','2015-12-30','2015-12-31','2016-01-01','2016-01-02','2016-01-03','2016-01-20','2016-01-21','2016-01-22','2016-02-01','2016-02-02','2016-02-20','2016-02-21','2016-02-14','2016-02-15','2016-03-07','2016-03-08','2016-03-09','2016-03-25','2016-03-26','2016-03-27']}

In [31]:
df_holiday=pd.DataFrame(holiday)

In [32]:
# Add holiday_flag
df['holiday_flag']= df.order_date.isin(df_holiday.holiday).astype(np.int8)

In [33]:
df.head()

fsn_id  order_date  Year  Month      order_id  order_item_id  \
0  ACCCX3S58G7B5F6P  2015-10-17  2015     10  3.419301e+15   3.419301e+15   
1  ACCCX3S58G7B5F6P  2015-10-19  2015     10  1.420831e+15   1.420831e+15   
2  ACCCX3S5AHMF55FV  2015-10-20  2015     10  2.421913e+15   2.421913e+15   
3  ACCCX3S5AHMF55FV  2015-10-14  2015     10  4.416592e+15   4.416592e+15   
4  ACCCX3S5AHMF55FV  2015-10-17  2015     10  4.419525e+15   4.419525e+15   

      gmv  units s1_fact.order_payment_type  sla  \
0  6400.0      1                        COD    5   
1  6900.0      1                        COD    7   
2  1990.0      1                        COD   10   
3  1690.0      1                    Prepaid    4   
4  1618.0      1                    Prepaid    6   

  product_analytic_super_category product_analytic_category  \
0                              CE           CameraAccessory   
1                              CE           CameraAccessory   
2                              CE           CameraAccessory   
3                              CE           CameraAccessory   
4                              CE           CameraAccessory   

  product_analytic_sub_category product_analytic_vertical  product_mrp  \
0               CameraAccessory              CameraTripod         7190   
1               CameraAccessory              CameraTripod         7190   
2               CameraAccessory              CameraTripod         2099   
3               CameraAccessory              CameraTripod         2099   
4               CameraAccessory              CameraTripod         2099   

   product_procurement_sla  Week_Number  pay_flag  holiday_flag  
0                        0           42         0             1  
1                        0           43         0             0  
2                        3           43         0             0  
3                        3           42         0             0  
4                        3           42         0             1

Understanding the Data Dictionary

In [34]:
df['s1_fact.order_payment_type'].astype('category').value_counts()

COD        1184291
Prepaid     448525
Name: s1_fact.order_payment_type, dtype: int64

In [35]:
df['product_analytic_super_category'].astype('category').value_counts()

CE    1632816
Name: product_analytic_super_category, dtype: int64

In [36]:
# dropping product_analytic_super_category column which is not having any variance
df=df.drop('product_analytic_super_category',axis=1)

In [37]:
df['product_analytic_category'].astype('category').value_counts()

EntertainmentSmall    935393
CameraAccessory       253581
GamingHardware        229994
GameCDDVD             113820
Camera                100028
Name: product_analytic_category, dtype: int64

In [38]:
df['product_analytic_sub_category'].astype('category').value_counts()

Speaker                535981
CameraAccessory        235717
GamingAccessory        201101
TVVideoSmall           142565
HomeAudio              124355
Game                   113723
AudioMP3Player         112544
Camera                 100028
GamingConsole           28893
CameraStorage           17864
AudioAccessory          11660
HomeTheatre              4245
AmplifierReceiver        4043
GameMembershipCards        97
Name: product_analytic_sub_category, dtype: int64

In [39]:
df['product_analytic_vertical'].astype('category').value_counts()

LaptopSpeaker              286602
MobileSpeaker              249379
AudioMP3Player             112544
PhysicalGame               104392
HomeAudioSpeaker            85298
GamingHeadset               62180
GamePad                     58939
DSLR                        56371
Flash                       48711
SelectorBox                 46116
RemoteControl               39583
CameraBattery               38754
Point & Shoot               37970
CameraTripod                37745
GamingMouse                 37350
VideoGlasses                36644
Lens                        35121
FMRadio                     28147
VideoPlayer                 20222
HandheldGamingConsole       19344
CameraStorageMemoryCard     17864
CameraBag                   17422
CameraBatteryCharger        15362
Binoculars                  14541
Filter                      12572
GamingConsole                9549
CodeInTheBoxGame             9331
GamingAccessoryKit           9323
Microphone                   9037
TVOutCableAcce

Filtering the dataframe  for camera accessory, home audio and gaming accessory.

In [40]:
subcategories=['CameraAccessory','GamingAccessory','HomeAudio ']

In [41]:
df_new=df[df.product_analytic_sub_category.isin(subcategories)]

In [42]:
df_new.shape

(436818, 18)

In [43]:
df_new.head()

fsn_id  order_date  Year  Month      order_id  order_item_id  \
0  ACCCX3S58G7B5F6P  2015-10-17  2015     10  3.419301e+15   3.419301e+15   
1  ACCCX3S58G7B5F6P  2015-10-19  2015     10  1.420831e+15   1.420831e+15   
2  ACCCX3S5AHMF55FV  2015-10-20  2015     10  2.421913e+15   2.421913e+15   
3  ACCCX3S5AHMF55FV  2015-10-14  2015     10  4.416592e+15   4.416592e+15   
4  ACCCX3S5AHMF55FV  2015-10-17  2015     10  4.419525e+15   4.419525e+15   

      gmv  units s1_fact.order_payment_type  sla product_analytic_category  \
0  6400.0      1                        COD    5           CameraAccessory   
1  6900.0      1                        COD    7           CameraAccessory   
2  1990.0      1                        COD   10           CameraAccessory   
3  1690.0      1                    Prepaid    4           CameraAccessory   
4  1618.0      1                    Prepaid    6           CameraAccessory   

  product_analytic_sub_category product_analytic_vertical  product_mrp  \
0               CameraAccessory              CameraTripod         7190   
1               CameraAccessory              CameraTripod         7190   
2               CameraAccessory              CameraTripod         2099   
3               CameraAccessory              CameraTripod         2099   
4               CameraAccessory              CameraTripod         2099   

   product_procurement_sla  Week_Number  pay_flag  holiday_flag  
0                        0           42         0             1  
1                        0           43         0             0  
2                        3           43         0             0  
3                        3           42         0             0  
4                        3           42         0             1

In [65]:
df_week=df_new.groupby('Week_Number').mean()
df_week.shape

(53, 11)

In [66]:
df_week.head(5)

Year  Month      order_id  order_item_id          gmv  \
Week_Number                                                            
1            2016.0    1.0  3.023925e+15   3.023925e+15   892.297322   
2            2016.0    1.0  3.002131e+15   3.002131e+15   993.378016   
3            2016.0    1.0  3.000041e+15   3.028014e+15  1115.139606   
4            2016.0    1.0  3.018744e+15   3.018744e+15  1159.971633   
5            2016.0    2.0  3.002875e+15   3.033218e+15   901.213021   

                units       sla  product_mrp  product_procurement_sla  \
Week_Number                                                             
1            1.025965  5.903065  2152.589757                 2.751553   
2            1.022615  6.235123  2218.573342                 2.583575   
3            1.031663  6.162810  2360.120094                 2.668009   
4            1.023582  6.351644  2056.874740                 2.606603   
5            1.027525  6.154610  1703.104472                 2.612402   

             pay_flag  holiday_flag  
Week_Number                          
1            0.000000      0.000000  
2            0.135313      0.000000  
3            0.000000      0.576206  
4            0.000000      0.000000  
5            0.116341      0.249258